internal/docstore/dynamodocstore/query_test.go

```diff
@@ -0,0 +1,372 @@
+			// The equality filter on the table's partition key allows us to query
+			// the table, but there is a better choice: a local index with a sort key
+			// that is mentioned in the query.
+			localIndexSortKey: "localS",
+			query: &driver.Query{Filters: []driver.Filter{
+				{[]string{"tableP"}, "=", 1},
+				{[]string{"localS"}, "<=", 1},
+			}},
+			want: &dynamodb.QueryInput{
+				IndexName:                aws.String("local"),
+				KeyConditionExpression:   aws.String("(#0 = :0) AND (#1 <= :1)"),
+				ExpressionAttributeNames: eans("tableP", "localS"),
+			},
+		},
+		{
+			desc: "equality filter on table partition, filters on local index and table sort",
+			// Given the choice of querying the table or a local index, prefer the table.
+			tableSortKey:      "tableS",
+			localIndexSortKey: "localS",
+			query: &driver.Query{Filters: []driver.Filter{
+				{[]string{"tableP"}, "=", 1},
+				{[]string{"localS"}, "<=", 1},
+				{[]string{"tableS"}, ">", 1},
+			}},
+			want: &dynamodb.QueryInput{
+				IndexName:                nil,
+				KeyConditionExpression:   aws.String("(#1 = :1) AND (#2 > :2)"),
+				FilterExpression:         aws.String("#0 <= :0"),
+				ExpressionAttributeNames: eans("localS", "tableP", "tableS"),
+			},
+		},
+		{
+			desc: "equality filter on other field with index",
+			// The query is the same as in "equality filter on other field," but now there
+			// is a global index with that field as partition key, so we can query it.
+			globalIndexPartitionKey: "other",
+			query:                   &driver.Query{Filters: []driver.Filter{{[]string{"other"}, "=", 1}}},
+			want: &dynamodb.QueryInput{
+				IndexName:                aws.String("global"),
+				KeyConditionExpression:   aws.String("#0 = :0"),
+				ExpressionAttributeNames: eans("other"),
+			},
+		},
+		{
+			desc: "equality filter on table partition, filter on global index sort",
+			// The equality filter on the table's partition key allows us to query
+			// the table, but there is a better choice: a global index with the same
+			// partition key and a sort key that is mentioned in the query.
+			globalIndexPartitionKey: "tableP",
+			globalIndexSortKey:      "globalS",
+			query: &driver.Query{Filters: []driver.Filter{
+				{[]string{"tableP"}, "=", 1},
+				{[]string{"globalS"}, "<=", 1},
+			}},
+			want: &dynamodb.QueryInput{
+				IndexName:                aws.String("global"),
+				KeyConditionExpression:   aws.String("(#0 = :0) AND (#1 <= :1)"),
+				ExpressionAttributeNames: eans("tableP", "globalS"),
+			},
+		},
+		{
+			desc: "equality filter on table partition, filter on global index sort, bad projection",
+			// Although there is a global index that matches the filters best, it doesn't
+			// have the necessary fields. So we query against the table.
+			globalIndexPartitionKey: "tableP",
+			globalIndexSortKey:      "globalS",
+			globalIndexFields:       []string{"other"},
+			query: &driver.Query{Filters: []driver.Filter{
```

Comment:
Consider adding a case that the query asks for some paths that the global index doesn't contain.

Reply:
Thanks, this case actually does that but not in a clear way. I fixed it.